[export] use tree_map for _flatten_dynamic_shapes

pianpwk opened the pull request:
Summary:
Fixing the implementation of `_flatten_dynamic_shapes()`, to follow how `_process_dynamic_shapes()` does it. The previous implementation would misinterpret some nested dynamic shapes specs, causing it to miss out on some shapes specs, for example with nested inputs/constant input tuples:

```
inputs = (
    (2, 1),
    (
        torch.randn(2, 1),
        torch.randn(2, 2),
        torch.randn(2, 3),
    )
)

dynamic_shapes = (
    (None, None),
    (
        None,
        None,
        None,
    )
)
```
This would get interpreted as 2 shapes specs for 2d and 3d tensors. Fixing so this doesn't happen.

Test Plan: Existing export tests

Differential Revision: D56894923

angelayi reviewed (approved):
can we add a test to make sure this change fixes the original issue? or does adding the assertion help us check the existing tests?

pianpwk committed:
[export] use tree_map for _flatten_dynamic_shapes (#125415)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/125415

Fixing the implementation of `_flatten_dynamic_shapes()`, to follow how `_process_dynamic_shapes()` does it. The previous implementation would misinterpret some nested dynamic shapes specs, causing it to miss out on some shapes specs, for example with nested inputs/constant input tuples:

```
inputs = (
    (2, 1),
    (
        torch.randn(2, 1),
        torch.randn(2, 2),
        torch.randn(2, 3),
    )
)

dynamic_shapes = (
    (None, None),
    (
        None,
        None,
        None,
    )
)
```
This would get interpreted as 2 shapes specs for 2d and 3d tensors. Fixing so this doesn't happen.

Test Plan: Existing export tests

Differential Revision: D56894923

pianpwk commented on torch/_export/non_strict_utils.py:
```diff
@@ -280,7 +270,16 @@ def make_constraints(
+    flat_dynamic_shapes = _flatten_dynamic_shapes(combined_args, dynamic_shapes)
+
+    # check number of shapes vs. number of inputs
+    num_placeholders = len([node for node in gm.graph.nodes if node.op == "placeholder"])
```
Thanks!

pianpwk commented:
> can we add a test to make sure this change fixes the original issue? or does adding the assertion help us check the existing tests?

Just added, thank you!
